File: app/utils/validators.py
# utils/validators.py
import re
import html

# Define a regex pattern for validating email formats (if needed in other areas)
EMAIL_REGEX = r'^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$'

def is_valid_id(value):
    """Check if a value is a valid positive integer (e.g., IDs)."""
    return isinstance(value, int) and value > 0

def is_valid_string(value, max_length=255):
    """Check if a value is a valid non-empty string within a maximum length."""
    return isinstance(value, str) and 0 < len(value) <= max_length

def sanitize_string(value):
    """Sanitize a string to escape HTML special characters."""
    if isinstance(value, str):
        return html.escape(value.strip())
    return ""

def is_valid_price(value):
    """Check if a value is a valid price (e.g., integer or float greater than zero)."""
    return isinstance(value, (int, float)) and value >= 0

def is_valid_quantity(value):
    """Check if a value is a valid quantity (non-negative integer)."""
    return isinstance(value, int) and value >= 0

def validate_warehouse_stock(warehouse_stock):
    """
    Validate and sanitize warehouse stock entries.
    Each entry should contain a valid warehouse_id and quantity.
    Returns a list of sanitized entries or raises a ValueError if validation fails.
    """
    sanitized_warehouse_stock = []
    for stock in warehouse_stock:
        warehouse_id = stock.get("warehouse_id")
        quantity = stock.get("quantity", 0)

        if is_valid_id(warehouse_id) and is_valid_quantity(quantity):
            sanitized_warehouse_stock.append({"warehouse_id": warehouse_id, "quantity": quantity})
        else:
            raise ValueError("Invalid warehouse stock entry")
    
    return sanitized_warehouse_stock

File: frontend/src/utils/validators.js
// utils/validators.js

// Validate that a string is not empty and does not exceed a max length
export const isValidString = (value, maxLength = 255) => {
    return typeof value === 'string' && value.trim().length > 0 && value.length <= maxLength;
};

// Validate that a number is a positive integer or zero
export const isValidQuantity = (value) => {
    const quantity = parseInt(value, 10);
    return Number.isInteger(quantity) && quantity >= 0;
};

// Validate price as a positive number
export const isValidPrice = (value) => {
    const price = parseFloat(value);
    return !isNaN(price) && price > 0;
};

// Escape HTML characters in a string to prevent XSS
export const sanitizeString = (value) => {
    return value.replace(/&/g, "&amp;")
                .replace(/</g, "&lt;")
                .replace(/>/g, "&gt;")
                .replace(/"/g, "&quot;")
                .replace(/'/g, "&#039;");
};

// Example validation for product data
export const validateProductData = (data) => {
    const { name, description, price, stock_quantity, category_id, subcategory_id } = data;
    return (
        isValidString(name) &&
        isValidString(description, 500) &&
        isValidPrice(price) &&
        isValidQuantity(stock_quantity) &&
        isValidQuantity(category_id) &&
        isValidQuantity(subcategory_id)
    );
};
